pyomo/contrib/iis/mis.py

```diff
@@ -0,0 +1,373 @@
+"""
+
+import logging
+import pyomo.environ as pyo
+
+from pyomo.core.plugins.transform.add_slack_vars import AddSlackVariables
+
+from pyomo.core.plugins.transform.hierarchy import IsomorphicTransformation
+
+from pyomo.common.modeling import unique_component_name
+from pyomo.common.collections import ComponentMap, ComponentSet
+
+from pyomo.opt import WriterFactory
+
+logger = logging.getLogger("pyomo.contrib.iis")
+logger.setLevel(logging.INFO)
+
+
+class _VariableBoundsAsConstraints(IsomorphicTransformation):
+    """Replace all variables bounds and domain information with constraints.
+
+    Leaves fixed Vars untouched (for now)
+    """
+
+    def _apply_to(self, instance, **kwds):
+
+        bound_constr_block_name = unique_component_name(instance, "_variable_bounds")
+        instance.add_component(bound_constr_block_name, pyo.Block())
+        bound_constr_block = instance.component(bound_constr_block_name)
+
+        for v in instance.component_data_objects(pyo.Var, descend_into=True):
+            if v.fixed:
+                continue
+            lb, ub = v.bounds
+            if lb is None and ub is None:
+                continue
+            var_name = v.getname(fully_qualified=True)
+            if lb is not None:
+                con_name = "lb_for_" + var_name
+                con = pyo.Constraint(expr=(lb, v, None))
+                bound_constr_block.add_component(con_name, con)
+            if ub is not None:
+                con_name = "ub_for_" + var_name
+                con = pyo.Constraint(expr=(None, v, ub))
+                bound_constr_block.add_component(con_name, con)
+
+            # now we deactivate the variable bounds / domain
+            v.domain = pyo.Reals
+            v.setlb(None)
+            v.setub(None)
+
+
+def compute_infeasibility_explanation(
+    model, solver, tee=False, tolerance=1e-8, logger=logger
+):
+    """
+    This function attempts to determine why a given model is infeasible. It deploys
+    two main algorithms:
+
+    1. Successfully relaxes the constraints of the problem, and reports to the user
+       some sets of constraints and variable bounds, which when relaxed, creates a
+       feasible model.
+    2. Uses the information collected from (1) to attempt to compute a Minimal
+       Infeasible System (MIS), which is a set of constraints and variable bounds
+       which appear to be in conflict with each other. It is minimal in the sense
+       that removing any single constraint or variable bound would result in a
+       feasible subsystem.
+
+    Args
+    ----
+        model: A pyomo block
+        solver: A pyomo solver object or a string for SolverFactory
+        tee (optional):  Display intermediate solves conducted (False)
+        tolerance (optional): The feasibility tolerance to use when declaring a
+            constraint feasible (1e-08)
+        logger:logging.Logger
+            A logger for messages. Uses pyomo.contrib.mis logger by default.
+
+    """
```

Comment:
It might be useful to return sets of names for each set of relaxed components, as well as the final minimal infeasible system. This doesn't need to hold up this PR.

Reply:
I'm going to put this suggestion in a comment.